File: windows-game/ViewModels/GameViewModel.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows;
using System.Windows.Threading;
using LongMarchWindows.Data;
using LongMarchWindows.Models;
using LongMarchWindows.Services;

namespace LongMarchWindows.ViewModels;

public class GameViewModel : INotifyPropertyChanged
{
    private readonly ProgressStore _store = new();
    private readonly DispatcherTimer _timer = new() { Interval = TimeSpan.FromMilliseconds(16) };

    private readonly HashSet<int> _completed = new();

    private float _playerX;
    private float _playerY;
    private int _playerStationId;
    private int _walkedUpTo;
    private bool _isWalking;
    private int? _selectedLevelId;
    private string _lockedTitle = string.Empty;
    private string _lockedDesc = string.Empty;

    // 行走动画中间状态
    private List<LevelInfo> _path = new();
    private int _segIndex;
    private LevelInfo _from = null!;
    private LevelInfo _to = null!;
    private DateTime _segStart;
    private DateTime _walkStart;
    private double _segDurMs;
    private int _targetId;
    private float _walkPhase;

    public GameViewModel()
    {
        var data = _store.Load();
        foreach (var id in data.CompletedLevels.Where(id => id is >= 1 and <= 10))
            _completed.Add(id);

        _playerStationId = Math.Clamp(data.PlayerStationId, 1, 10);
        _walkedUpTo = _playerStationId;

        var start = Levels.First(l => l.Id == _playerStationId);
        _playerX = start.X;
        _playerY = start.Y - PlayerYOffset;

        _timer.Tick += OnTimerTick;
    }

    public const float PlayerYOffset = 25f;

    public IReadOnlyList<LevelInfo> Levels => GameData.Levels;

    public event PropertyChangedEventHandler? PropertyChanged;

    public float PlayerX
    {
        get => _playerX;
        private set => SetField(ref _playerX, value);
    }

    public float PlayerY
    {
        get => _playerY;
        private set => SetField(ref _playerY, value);
    }

    public bool IsWalking
    {
        get => _isWalking;
        private set => SetField(ref _isWalking, value);
    }

    public int WalkedUpTo
    {
        get => _walkedUpTo;
        private set => SetField(ref _walkedUpTo, value);
    }

    public float WalkPhase
    {
        get => _walkPhase;
        private set => SetField(ref _walkPhase, value);
    }

    public int? SelectedLevelId
    {
        get => _selectedLevelId;
        private set => SetField(ref _selectedLevelId, value);
    }

    public string LockedTitle
    {
        get => _lockedTitle;
        private set => SetField(ref _lockedTitle, value);
    }

    public string LockedDesc
    {
        get => _lockedDesc;
        private set => SetField(ref _lockedDesc, value);
    }

    public Visibility LockedOverlayVisibility =>
        string.IsNullOrEmpty(_lockedTitle) ? Visibility.Collapsed : Visibility.Visible;

    public int CurrentLevel => Math.Min(10, (_completed.Count == 0 ? 0 : _completed.Max()) + 1);

    public string ProgressText =>
        $"当前进度：第 {CurrentLevel} / {Levels.Count} 关  |  点击关卡节点继续征程";

    public LevelStatus GetStatus(int id)
    {
        if (_completed.Contains(id)) return LevelStatus.Completed;
        if (id <= CurrentLevel) return LevelStatus.Available;
        return LevelStatus.Locked;
    }

    public int GetStars(int id) => _completed.Contains(id) ? 3 : 0;

    public void ClickStation(int id)
    {
        SelectedLevelId = id;
        var level = Levels.First(l => l.Id == id);

        if (GetStatus(id) == LevelStatus.Locked)
        {
            LockedTitle = level.MapLabel;
            LockedDesc = level.ShortDesc;
            RaiseLockedOverlayChanged();
            return;
        }

        DismissLockedInfo();
        StartWalk(id);
    }

    public void DismissLockedInfo()
    {
        if (string.IsNullOrEmpty(_lockedTitle)) return;
        LockedTitle = string.Empty;
        LockedDesc = string.Empty;
        RaiseLockedOverlayChanged();
    }

    public void Reset()
    {
        if (_isWalking)
        {
            _timer.Stop();
            _isWalking = false;
        }

        _completed.Clear();
        _playerStationId = 1;
        _walkedUpTo = 1;
        SelectedLevelId = null;
        WalkPhase = 0;
        DismissLockedInfo();

        var start = Levels.First(l => l.Id == 1);
        PlayerX = start.X;
        PlayerY = start.Y - PlayerYOffset;

        OnPropertyChanged(nameof(CurrentLevel));
        OnPropertyChanged(nameof(ProgressText));
        OnPropertyChanged(nameof(WalkedUpTo));
        Save();
    }

    private void RaiseLockedOverlayChanged()
    {
        OnPropertyChanged(nameof(LockedTitle));
        OnPropertyChanged(nameof(LockedDesc));
        OnPropertyChanged(nameof(LockedOverlayVisibility));
    }

    private void StartWalk(int targetId)
    {
        if (_isWalking) return;

        _path = BuildPath(_playerStationId, targetId);
        _targetId = targetId;

        if (_path.Count < 2)
        {
            CompleteArrival(targetId);
            return;
        }

        _segIndex = 0;
        _walkStart = DateTime.Now;
        IsWalking = true;
        BeginSegment();
    }

    private List<LevelInfo> BuildPath(int fromId, int toId)
    {
        int start = Math.Min(fromId, toId);
        int end = Math.Max(fromId, toId);

        var list = new List<LevelInfo>();
        for (int i = start; i <= end; i++)
            list.Add(Levels.First(l => l.Id == i));

        if (toId < fromId) list.Reverse();
        return list;
    }

    private void BeginSegment()
    {
        _from = _path[_segIndex];
        _to = _path[_segIndex + 1];

        float dx = _to.X - _from.X;
        float dy = (_to.Y - PlayerYOffset) - (_from.Y - PlayerYOffset);
        float dist = MathF.Sqrt(dx * dx + dy * dy);
        _segDurMs = Math.Max(dist / 80f, 0.3f) * 1000;
        _segStart = DateTime.Now;
        _timer.Start();
    }

    private void OnTimerTick(object? sender, EventArgs e)
    {
        WalkPhase = (float)(((DateTime.Now - _walkStart).TotalMilliseconds % 400) / 400.0);

        double elapsed = (DateTime.Now - _segStart).TotalMilliseconds;
        double raw = Math.Min(elapsed / _segDurMs, 1.0);
        double t = raw < 0.5 ? 2 * raw * raw : -1 + (4 - 2 * raw) * raw;

        float fromX = _from.X;
        float fromY = _from.Y - PlayerYOffset;
        float toX = _to.X;
        float toY = _to.Y - PlayerYOffset;

        PlayerX = fromX + (toX - fromX) * (float)t;
        PlayerY = fromY + (toY - fromY) * (float)t;

        if (raw < 1.0) return;

        _segIndex++;
        WalkedUpTo = _to.Id;

        if (_segIndex >= _path.Count - 1)
        {
            _timer.Stop();
            IsWalking = false;
            CompleteArrival(_targetId);
        }
        else
        {
            BeginSegment();
        }
    }

    private void CompleteArrival(int id)
    {
        var level = Levels.First(l => l.Id == id);
        PlayerX = level.X;
        PlayerY = level.Y - PlayerYOffset;
        _playerStationId = id;
        WalkedUpTo = id;
        WalkPhase = 0;

        bool unlockedNew = false;
        if (!_completed.Contains(id))
        {
            _completed.Add(id);
            unlockedNew = true;
        }

        if (unlockedNew)
        {
            OnPropertyChanged(nameof(CurrentLevel));
            OnPropertyChanged(nameof(ProgressText));
        }

        Save();
    }

    private void Save()
    {
        _store.Save(new ProgressStore.ProgressData(_completed.OrderBy(x => x).ToList(), _playerStationId));
    }

    private void SetField<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return;
        field = value;
        OnPropertyChanged(name);
    }

    private void OnPropertyChanged([CallerMemberName] string? name = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}

File: android-app/app/src/main/java/com/longmarch/game/ui/GameViewModel.kt
package com.longmarch.game.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.longmarch.game.data.GameData
import com.longmarch.game.data.LevelInfo
import com.longmarch.game.data.LevelStatus
import com.longmarch.game.data.ProgressData
import com.longmarch.game.data.ProgressRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

data class GameUiState(
    val playerX: Float = 0f,
    val playerY: Float = 0f,
    val isWalking: Boolean = false,
    val walkPhase: Float = 0f,
    val walkedUpTo: Int = 1,
    val selectedLevelId: Int? = null,
    val lockedTitle: String? = null,
    val lockedDesc: String? = null,
    val completedLevels: Set<Int> = emptySet(),
    val playerStationId: Int = 1,
) {
    val currentLevel: Int get() = min(10, (completedLevels.maxOrNull() ?: 0) + 1)

    val progressText: String
        get() = "当前进度：第 $currentLevel / ${GameData.levels.size} 关  |  点击关卡节点继续征程"

    val isLockedInfoVisible: Boolean get() = lockedTitle != null

    fun status(id: Int): LevelStatus = when {
        id in completedLevels -> LevelStatus.COMPLETED
        id <= currentLevel -> LevelStatus.AVAILABLE
        else -> LevelStatus.LOCKED
    }

    fun stars(id: Int): Int = if (id in completedLevels) 3 else 0
}

class GameViewModel(application: Application) : AndroidViewModel(application) {

    companion object {
        const val PlayerYOffset = 25f
    }

    private val repository = ProgressRepository(application)

    private val _uiState = MutableStateFlow(GameUiState())
    val uiState: StateFlow<GameUiState> = _uiState.asStateFlow()

    private var walkJob: Job? = null

    init {
        viewModelScope.launch {
            val progress = repository.progress.first()
            val start = GameData.levels.first { it.id == progress.playerStationId }
            _uiState.update {
                it.copy(
                    playerX = start.x,
                    playerY = start.y - PlayerYOffset,
                    playerStationId = progress.playerStationId,
                    walkedUpTo = progress.playerStationId,
                    completedLevels = progress.completedLevels,
                )
            }
        }
    }

    fun clickStation(id: Int) {
        val level = GameData.levels.first { it.id == id }
        _uiState.update { it.copy(selectedLevelId = id) }

        if (_uiState.value.status(id) == LevelStatus.LOCKED) {
            _uiState.update { it.copy(lockedTitle = level.mapLabel, lockedDesc = level.shortDesc) }
            return
        }

        dismissLockedInfo()
        startWalk(id)
    }

    fun dismissLockedInfo() {
        if (!_uiState.value.isLockedInfoVisible) return
        _uiState.update { it.copy(lockedTitle = null, lockedDesc = null) }
    }

    fun reset() {
        walkJob?.cancel()
        val start = GameData.levels.first { it.id == 1 }
        _uiState.update {
            it.copy(
                isWalking = false,
                walkPhase = 0f,
                playerStationId = 1,
                walkedUpTo = 1,
                selectedLevelId = null,
                lockedTitle = null,
                lockedDesc = null,
                completedLevels = emptySet(),
                playerX = start.x,
                playerY = start.y - PlayerYOffset,
            )
        }
        save()
    }

    // ==================================================================
    // 行走动画（与 Windows 版 DispatcherTimer 逻辑一致）
    // ==================================================================

    private fun startWalk(targetId: Int) {
        if (_uiState.value.isWalking) return

        val path = buildPath(_uiState.value.playerStationId, targetId)
        if (path.size < 2) {
            completeArrival(targetId)
            return
        }

        walkJob = viewModelScope.launch {
            val walkStart = System.currentTimeMillis()
            _uiState.update { it.copy(isWalking = true) }

            for (segIndex in 0 until path.size - 1) {
                val from = path[segIndex]
                val to = path[segIndex + 1]
                val dx = to.x - from.x
                val dy = (to.y - PlayerYOffset) - (from.y - PlayerYOffset)
                val dist = sqrt(dx * dx + dy * dy)
                val segDurMs = max(dist / 80f, 0.3f) * 1000f
                val segStart = System.currentTimeMillis()

                while (isActive) {
                    val now = System.currentTimeMillis()
                    val phase = ((now - walkStart).toFloat() % 400f) / 400f
                    val elapsed = (now - segStart).toFloat()
                    val raw = (elapsed / segDurMs).coerceIn(0f, 1f)
                    val t = if (raw < 0.5f) 2f * raw * raw else -1f + (4f - 2f * raw) * raw
                    val px = from.x + dx * t
                    val py = (from.y - PlayerYOffset) + dy * t

                    _uiState.update { it.copy(playerX = px, playerY = py, walkPhase = phase) }
                    if (raw >= 1f) break
                    delay(16)
                }

                _uiState.update { it.copy(walkedUpTo = to.id) }
            }

            completeArrival(targetId)
        }
    }

    private fun buildPath(fromId: Int, toId: Int): List<LevelInfo> {
        val start = min(fromId, toId)
        val end = max(fromId, toId)
        val list = (start..end).map { id -> GameData.levels.first { it.id == id } }
        return if (toId < fromId) list.reversed() else list
    }

    private fun completeArrival(id: Int) {
        val level = GameData.levels.first { it.id == id }
        _uiState.update { state ->
            state.copy(
                playerX = level.x,
                playerY = level.y - PlayerYOffset,
                playerStationId = id,
                walkedUpTo = id,
                walkPhase = 0f,
                isWalking = false,
                completedLevels = state.completedLevels + id,
            )
        }
        save()
    }

    private fun save() {
        val s = _uiState.value
        viewModelScope.launch {
            repository.save(ProgressData(s.completedLevels, s.playerStationId))
        }
    }
}
